### PR TITLE
Manipulation demo: Add streamlit interface

### PR DESCRIPTION
**Pull request by knicked**
## Purpose

This PR enhances the manipulation demo with a streamlit interface, based on the implementation of `debugging_assistant.py`.

## Proposed changes
- Add streamlit interface
- Display all tool calls and tool call history

## Testing

```bash
streamlit run examples/manipulation-demo-streamlit.py
```

### CODE REVIEW

File: docs/demos/manipulation.md [hunk 1]
```diff
@@ -50,12 +50,16 @@ This demo showcases the capabilities of RAI (Robotec AI) in performing manipulat
...
+2. In the second terminal, run the cli prompt or streamlit interface:
 
    ```shell
    python examples/manipulation-demo.py
    ```
 
+   ```shell
+   streamlit run examples/manipulation-demo-streamlit.py
+   ```
+
```
Comment (maciejmajek) [inline docs/demos/manipulation.md]: Please add a few words of explanation for the `cli prompt` and `streamlit interface`.
e.g. That the cli version provides a simple script showing the rai api that could be used in other places whereas the streamlit is better for trying the demo on it's own

Comment (knicked) [reply docs/demos/manipulation.md]: Done.

File: examples/manipulation-demo-streamlit.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                st.code(msg.content, language="json")
+
+    if prompt:
+        st.session_state.messages.append(HumanMessage(content=prompt))
+        st.chat_message("user").write(prompt)
+        with st.chat_message("assistant"):
+            st_callback = get_streamlit_cb(st.container())
+            streamlit_invoke(
+                st.session_state["graph"], st.session_state.messages, [st_callback]
+            )
```
Comment (boczekbartek) [inline examples/manipulation-demo-streamlit.py:70]: @knicked 
Not sure if this should be addressed under this PR, but I noticed that camera image is not shown in streamlit:
image

File: docs/demos/manipulation.md [hunk 3]
```diff
@@ -50,12 +50,20 @@ This demo showcases the capabilities of RAI in performing manipulation tasks usi
...
    ```shell
    python examples/manipulation-demo.py
    ```
 
+   This script will prompt you to enter a natural language command in the terminal. It also serves as an example of how to use the RAI API for your own applications.
+
+   If you just want to try the demo on its own with a more user-friendly interface, you can use the Streamlit version:
+
+   ```shell
+   streamlit run examples/manipulation-demo-streamlit.py
```
Comment (boczekbartek) [inline docs/demos/manipulation.md]: I would make streamlit the default. We can keep the terminal one for now, but I would prioritize streamlit.

**Review by boczekbartek — APPROVED** [review]
@knicked 
I tested this PR in combination with https://github.com/RobotecAI/rai/pull/341
I used multiple prompts such as:
- move corn to the carrot
- move boxes close to each other
- grab apple
And they all succeeded (with some coliders issues, but it's not the topic of this PR).

2 nitpick comments left ;)

File: examples/manipulation-demo-streamlit.py (hunk 2 above)
Comment (knicked) [reply examples/manipulation-demo-streamlit.py:70]: This PR uses the currently existing streamlit integration with the llm agent, which apparently does not support displaying images. I will add this functionality in a separate PR.

File: docs/demos/manipulation.md (hunk 3 above)
Comment (knicked) [reply docs/demos/manipulation.md]: I updated the docs to address that :+1:

File: examples/manipulation-demo-streamlit.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    Before starting the task, make sure to grab the camera image to understand the environment.
+    """
+
+    llm = get_llm_model(model_type="complex_model", streaming=True)
+    agent = create_conversational_agent(
+        llm=llm,
+        tools=tools,
+        system_prompt=system_prompt,
+    )
+    return agent
```
Comment (boczekbartek) [inline examples/manipulation-demo-streamlit.py]: This part is copied from [here](https://github.com/RobotecAI/rai/blob/47e74fe34d141274783ceb8a535c420c2bcc6758/examples/manipulation-demo.py#L24-L58). Please refactor

**Review by boczekbartek — CHANGES_REQUESTED** [review]

File: examples/manipulation-demo-streamlit.py (hunk 4 above)
Comment (knicked) [reply examples/manipulation-demo-streamlit.py]: Done.

Comment (boczekbartek) [reply examples/manipulation-demo-streamlit.py]: for some reason after the change (or rebase) streamlit doesn't work for me:
image

Comment (boczekbartek) [reply examples/manipulation-demo-streamlit.py]:
```
(rai-py3.12) robo-pc-005 ➜  12_rai git:(kd/manipulation_streamlit) ✗ streamlit run examples/manipulation-demo-streamlit.py 

  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://172.16.0.47:8501

FutureWarning: Importing from timm.models.layers is deprecated, please import via timm.layers
2024-12-20 15:51:02 robo-pc-005 rai.utils.model_initialization[265091] INFO Initializing complex_model: Vendor: openai, Model: gpt-4o-2024-08-06
2024-12-20 15:51:02 robo-pc-005 rai.agents.conversational_agent[265091] INFO Creating state based agent
2024-12-20 15:51:02 robo-pc-005 rai.agents.conversational_agent[265091] INFO State based agent created
2024-12-20 15:51:06 robo-pc-005 rai.agents.conversational_agent[265091] INFO Running thinker
2024-12-20 15:51:10 robo-pc-005 httpx[265091] INFO HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
```

Comment (boczekbartek) [reply examples/manipulation-demo-streamlit.py]: I tried `streamlit run src/rai/rai/tools/debugging_assistant.py` and it works well. So I think it's sth wrong with this PR

Comment (knicked) [reply examples/manipulation-demo-streamlit.py]: I can confirm it's not working correctly. The agent does call tools and respond, but the last message seems to only be displayed after the user types in another message. I'm looking into this now.

Comment (knicked) [reply examples/manipulation-demo-streamlit.py]: The issue should be solved now. I made a small mistake when refactoring the duplicated code and missed it when testing, thanks for noticing!

Comment (boczekbartek) [reply examples/manipulation-demo-streamlit.py]: now it works great! thanks

**Review by boczekbartek — APPROVED** [review]
Looks good!

**Pull request merged**